### PR TITLE
UCC CI github actions (DPU server)

### PR DESCRIPTION
**Pull request by B-a-S**
Added GA CI support for branch v0.4.x
Ex.:
https://github.com/B-a-S/ucc/actions/runs/4472192413

### CODE REVIEW

File: .github/workflows/dpu-server.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+name: DPU server CI
```
Comment (artemry-nv) [inline .github/workflows/dpu-server.yaml:1]: I suppose `UCC service` is more appropriate name for this product, isn't it?

Comment (artemry-nv) [reply .github/workflows/dpu-server.yaml:1]: If yes, please reflect in the filenames as well

File: .github/workflows/dpu-server.yaml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      run: /start build_arm
+    - name: Running DPU server
+      run: /start dpu_server
+  run_dpu:
+    needs: [build_dpu, build_x86]
+    runs-on: [self-hosted, linux , x64]  
+    steps:
+    - name: Running DPU server
+      run: /start run_dpu
+  run_omb:
```
Comment (artemry-nv) [inline .github/workflows/dpu-server.yaml:36]: Maybe to rename to run_test?

File: .github/workflows/dpu-server.yaml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      run: /start build_x86
+  build_dpu:
+    needs: [deployment]
+    runs-on: [self-hosted, linux , x64]  
+    steps:
+    - name: Building on the DPU
+      run: /start build_arm
+    - name: Running DPU server
+      run: /start dpu_server
+  run_dpu:
```
Comment (artemry-nv) [inline .github/workflows/dpu-server.yaml:30]: Maybe to rename to deploy_ucc_service?

File: .github/workflows/dpu-server.yaml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  run_omb:
+    needs: [build_dpu, build_x86]
+    runs-on: [self-hosted, linux , x64]  
+    steps:
+    - name: Running tests
+      run: /start run_omb
+  clean:
+    if: ${{ always() }}
+    needs: [build_dpu, build_x86, run_omb, run_dpu]
+    runs-on: [self-hosted, linux , x64]
```
Comment (artemry-nv) [inline .github/workflows/dpu-server.yaml]: Odd space

File: contrib/dpu_daemon/Makefile [hunk 5]
```diff
@@ -7,10 +7,10 @@
 # $HEADER$
 #
 
-MPI_DIR=$(OMPI_DIR)
-#UCX_DIR=$(UCX_DIR)
+#MPI_DIR=$(HPCX_MPI_DIR)
+#UCX_DIR=$(HPCX_UCX_DIR)
 #UCC_DIR=/global/scratch/users/sourav/build-arm/ucc
-UCC_SRC=/global/scratch/users/tomislavj/dpu-dev/ucc
+#UCC_SRC=/global/home/users/sourav/workspace/ucc
```
Comment (artemry-nv) [inline contrib/dpu_daemon/Makefile]: Do we need this commented code?

File: contrib/dpu_daemon/scripts/build_arm.sh [hunk 6]
```diff
@@ -2,63 +2,60 @@
...
-OMB_DIR="$BUILD_DIR/omb"
-OMB_URL='https://github.com/paklui/osu-micro-benchmarks.git'
-OMB_BRANCH='master'
+DIR=$(dirname `realpath $0`)
+pushd ${DIR}
+. ./config.sh
 
 echo "#### WORK DIR  : $WORK_DIR  ####"
 echo "#### BUILD DIR : $BUILD_DIR ####"
+mkdir -p ${WORK_DIR}
```
Comment (artemry-nv) [inline contrib/dpu_daemon/scripts/build_arm.sh]: Please fix formatting and shellcheck issues

File: contrib/dpu_daemon/scripts/config.sh [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+export OMB_URL='https://github.com/paklui/osu-micro-benchmarks.git'
+export OMB_BRANCH=${OMB_BRANCH:-master}
+
+completed (){
+    if [ ${MULTI_BUILD} -eq 1 ]; then
+        touch ${WORK_DIR}/TMP/${1}
+    fi
+}
+
+waiting (){
```
Comment (artemry-nv) [inline contrib/dpu_daemon/scripts/config.sh]: Maybe rename to something like `wait_for` (infinitive is usually used in function names)?

File: contrib/dpu_daemon/scripts/build_arm.sh [hunk 8]
```diff
@@ -2,63 +2,60 @@
...
+        mkdir -p $MPI_DIR
+        mkdir -p $OMB_DIR
 cd $WORK_DIR
 
+waiting downloads
+
 if [ $download -eq 1 ]; then
     if [ ! -d $UCX_SRC ]; then
-        echo "#### Downloading UCX from $UCX_URL:$UCX_BRANCH ####"
+        echo "${GH_FOLD}#### Downloading UCX from $UCX_URL:$UCX_BRANCH ####"
```
Comment (artemry-nv) [inline contrib/dpu_daemon/scripts/build_arm.sh]: Suggest to add a echo wrapper function with this GH macro to remove duplicated code

File: contrib/dpu_daemon/scripts/run_dpu.sh [hunk 9]
```diff
@@ -32,3 +30,7 @@ cmd="$MPI_DIR/bin/mpirun --np ${NP} \
 echo $cmd
 eval "$cmd"
 
+if [ $? -eq 130 ];
```
Comment (artemry-nv) [inline contrib/dpu_daemon/scripts/run_dpu.sh]: What does 130 mean here?

File: contrib/dpu_daemon/scripts/run_omb.sh [hunk 10]
```diff
@@ -1,25 +1,28 @@
...
 BSZ=$((4))
 ESZ=$((128 * 1024 * 1024))
 MEMSZ=$((16 * 1024 * 1024 * 1024))
 ITER=200
 WARM=20
-
-HOSTS=${1}
+MLX_DEV="mlx5_0:1"
+HOSTS=2
+#${1}
```
Comment (artemry-nv) [inline contrib/dpu_daemon/scripts/run_omb.sh]: Do we need this commented code?

Comment (artemry-nv) [reply contrib/dpu_daemon/scripts/run_omb.sh]: Check other places

File: .github/workflows/dpu-server.yaml (hunk 1 above)
Comment (B-a-S) [reply .github/workflows/dpu-server.yaml:1]: The name of the file is dpu_server. So, it's ok.

File: .github/workflows/dpu-server.yaml (hunk 3 above)
Comment (B-a-S) [reply .github/workflows/dpu-server.yaml:30]: This step  just runs dpu_server app

File: contrib/dpu_daemon/Makefile (hunk 5 above)
Comment (B-a-S) [reply contrib/dpu_daemon/Makefile]: I just commented a legacy and hardcoded code

File: contrib/dpu_daemon/scripts/build_arm.sh (hunk 8 above)
Comment (B-a-S) [reply contrib/dpu_daemon/scripts/build_arm.sh]: Fixed

File: contrib/dpu_daemon/scripts/config.sh (hunk 7 above)
Comment (B-a-S) [reply contrib/dpu_daemon/scripts/config.sh]: Renamed

File: contrib/dpu_daemon/scripts/run_dpu.sh (hunk 9 above)
Comment (B-a-S) [reply contrib/dpu_daemon/scripts/run_dpu.sh]: Added a comment: # No error when dpu server killed by kill -2 dpu_master

File: contrib/dpu_daemon/scripts/run_omb.sh (hunk 10 above)
Comment (B-a-S) [reply contrib/dpu_daemon/scripts/run_omb.sh]: Removed

File: contrib/dpu_daemon/scripts/build_arm.sh (hunk 6 above)
Comment (B-a-S) [reply contrib/dpu_daemon/scripts/build_arm.sh]: Fixed

File: .github/workflows/dpu-server.yaml (hunk 2 above)
Comment (B-a-S) [reply .github/workflows/dpu-server.yaml:36]: This step doesn't run the tests.

File: .github/workflows/dpu-server.yaml (hunk 4 above)
Comment (B-a-S) [reply .github/workflows/dpu-server.yaml]: Done

File: .github/workflows/dpu-server.yaml [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+on:
+  workflow_dispatch:
+    branches: [ "v0.4.x"]
+  push:
+    branches: [ "v0.4.x"]
+  pull_request: 
+    branches: [ "v0.4.x"]
+jobs:
+  deployment:
+    runs-on: [self-hosted, linux , x64]
```
Comment (artemry-nv) [inline .github/workflows/dpu-server.yaml]: Odd space symbol

Comment (artemry-nv) [reply .github/workflows/dpu-server.yaml]: Check other similar places

File: .github/workflows/dpu-server.yaml (hunk 4 above)
Comment (artemry-nv) [reply .github/workflows/dpu-server.yaml]: Seems not yet fixed

File: contrib/dpu_daemon/Makefile (hunk 5 above)
Comment (artemry-nv) [reply contrib/dpu_daemon/Makefile]: Let's remove all useless commented code

File: contrib/dpu_daemon/scripts/build_arm.sh [hunk 12]
```diff
@@ -1,64 +1,61 @@
...
-WORK_DIR='/global/scratch/users/dpu/deploy'
-BUILD_DIR="$WORK_DIR/build-$arch"
-
-UCX_SRC="$WORK_DIR/ucx"
-UCX_DIR="$BUILD_DIR/ucx"
-UCX_URL='https://github.com/openucx/ucx.git'
-UCX_BRANCH='master'
+DIR=$(dirname `realpath $0`)
+pushd ${DIR}
+. ./config.sh
```
Comment (artemry-nv) [inline contrib/dpu_daemon/scripts/build_arm.sh]: Have you checked all .sh files with shellcheck?

File: contrib/dpu_daemon/scripts/build_arm.sh (hunk 8 above)
Comment (artemry-nv) [reply contrib/dpu_daemon/scripts/build_arm.sh]: Don't see the wrapper function

File: contrib/dpu_daemon/scripts/build_x86.sh [hunk 13]
```diff
@@ -1,122 +1,121 @@
-#!/bin/bash
-set -e
-
+#!/bin/bash -eE
 arch=x86
```
Comment (artemry-nv) [inline contrib/dpu_daemon/scripts/build_x86.sh]: Consider using `set -o pipefail` and `trap ERR`

File: contrib/dpu_daemon/scripts/config.sh [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+export MPI_DIR="$BUILD_DIR/ompi"
+export MPI_URL='https://github.com/open-mpi/ompi.git'
+export MPI_BRANCH=${MPI_BRANCH:-v5.0.x}
+
+export OMB_SRC="$WORK_DIR/omb"
+export OMB_DIR="$BUILD_DIR/omb"
+export OMB_URL='https://github.com/paklui/osu-micro-benchmarks.git'
+export OMB_BRANCH=${OMB_BRANCH:-master}
+
+completed (){
```
Comment (artemry-nv) [inline contrib/dpu_daemon/scripts/config.sh]: `completed` as a function looks a bit non-standard - maybe rename it to something more convenient?

File: .github/workflows/dpu-server.yaml (hunk 11 above)
Comment (B-a-S) [reply .github/workflows/dpu-server.yaml]: fixed

File: contrib/dpu_daemon/scripts/build_x86.sh (hunk 13 above)
Comment (B-a-S) [reply contrib/dpu_daemon/scripts/build_x86.sh]: There are no pipes

File: contrib/dpu_daemon/scripts/build_x86.sh [hunk 15]
```diff
@@ -1,122 +1,121 @@
...
+    echo "${GH_FOLD}#### Building MPI ####"
 	cd $MPI_SRC
 	./autogen.pl
 	mkdir -p $MPI_DIR
 	cd $MPI_DIR
+    
+        completed configure_mpi
 
-    echo "#### Confguring MPI ####"
+    echo "${GH_FOLD}#### Confguring MPI ####"
```
Comment (artemry-nv) [inline contrib/dpu_daemon/scripts/build_x86.sh]: Misprint

File: contrib/dpu_daemon/scripts/build_arm.sh [hunk 16]
```diff
@@ -1,64 +1,61 @@
-#!/bin/bash
-set -e
-
+#!/bin/bash -eE
 arch=arm
```
Comment (artemry-nv) [inline contrib/dpu_daemon/scripts/build_arm.sh]: Please combine build scripts into one

File: contrib/dpu_daemon/scripts/build_arm.sh [hunk 17]
```diff
@@ -1,64 +1,61 @@
...
-echo "#### WORK DIR  : $WORK_DIR  ####"
-echo "#### BUILD DIR : $BUILD_DIR ####"
-cd $WORK_DIR
+wait_for downloads
 
 if [ $download -eq 1 ]; then
     if [ ! -d $UCX_SRC ]; then
-        echo "#### Downloading UCX from $UCX_URL:$UCX_BRANCH ####"
+        echo "${GH_FOLD}#### Downloading UCX from $UCX_URL:$UCX_BRANCH ####"
         git clone -b $UCX_BRANCH $UCX_URL ucx
```
Comment (artemry-nv) [inline contrib/dpu_daemon/scripts/build_arm.sh]: Move clone into common preliminary step

**Review by artemry-nv — CHANGES_REQUESTED** [review]

File: contrib/dpu_daemon/scripts/config.sh [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+  [${PKG_INDEX[1]}, BRANCH]="${UCC_BRANCH:-v0.4.x}"
+  [${PKG_INDEX[1]}, SRC]="${WORK_DIR}/${PKG_INDEX[1]}"
+  [${PKG_INDEX[1]}, BUILD]="${BUILD_DIR}/${PKG_INDEX[1]}"
+  #OMPI
+  [${PKG_INDEX[2]}, URL]='https://github.com/open-mpi/ompi.git'
+  [${PKG_INDEX[2]}, BRANCH]="${MPI_BRANCH:-v5.0.x}"
+  [${PKG_INDEX[2]}, SRC]="${WORK_DIR}/${PKG_INDEX[2]}"
+  [${PKG_INDEX[2]}, BUILD]="${BUILD_DIR}/${PKG_INDEX[2]}"
+  #OMB
+  [${PKG_INDEX[3]}, URL]='https://github.com/paklui/osu-micro-benchmarks.git'
```
Comment (artemry-nv) [inline contrib/dpu_daemon/scripts/config.sh:51]: Strange OSU bench repo - as far as I see here're the official src http://mvapich.cse.ohio-state.edu/benchmarks/

File: contrib/dpu_daemon/scripts/config.sh [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+)
+
+#= List of packages
+declare -a PKG_INDEX=(${BUILD_INDEX[@]:0:4})
+
+#= Configurations for 'git clone'
+declare -A GIT=(
+  #UCX
+  [${PKG_INDEX[0]}, URL]='https://github.com/openucx/ucx.git'
+  [${PKG_INDEX[0]}, BRANCH]="${UCX_BRANCH:-v1.14.x}"
```
Comment (artemry-nv) [inline contrib/dpu_daemon/scripts/config.sh:37]: UCX v1.15.x is available now - should we switch?

File: contrib/dpu_daemon/scripts/run.sh [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+    DPU)
+        cd "${GIT[UCC, SRC]}"/contrib/dpu_daemon
+        export OMPI_DIR="${GIT[OMPI, BUILD]}"
+        export UCX_DIR="${GIT[UCX, BUILD]}"
+        export UCC_DIR="${GIT[UCC, BUILD]}"
+        export UCC_SRC="${GIT[UCC, SRC]}"
+        make BUILD_DIR="${BUILD_DIR}"
+        ;;
+    *)
+        echo "Do nothing, no such option to build !!!"
```
Comment (artemry-nv) [inline contrib/dpu_daemon/scripts/run.sh:122]: Should we exit with error here?

File: .github/workflows/dpu-server.yaml [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+      run: /start deploy
+  build_x86:
+    needs: [deployment]
+    runs-on: [self-hosted, linux, x64]  
+    steps:
+    - name: Building on the HOST
+      run: /start build_x86
+  build_dpu:
+    needs: [deployment]
+    runs-on: [self-hosted, linux, x64]
```
Comment (artemry-nv) [inline .github/workflows/dpu-server.yaml:26]: Do we build dpu part on x64?

**Review by artemry-nv — APPROVED** [review]

File: contrib/dpu_daemon/scripts/config.sh (hunk 18 above)
Comment (B-a-S) [reply contrib/dpu_daemon/scripts/config.sh:51]: It'll be done in the next PR

File: contrib/dpu_daemon/scripts/config.sh (hunk 19 above)
Comment (B-a-S) [reply contrib/dpu_daemon/scripts/config.sh:37]: We still have to use v1.14 with UCC branch v0.4.x

File: .github/workflows/dpu-server.yaml (hunk 21 above)
Comment (B-a-S) [reply .github/workflows/dpu-server.yaml:26]: We have only x84 runners, but it uses dpu as well

File: contrib/dpu_daemon/scripts/run.sh (hunk 20 above)
Comment (B-a-S) [reply contrib/dpu_daemon/scripts/run.sh:122]: Will be done in the following PR

**Pull request merged**